### DISCUSSION TITLE
Removing mandatory `break` from switch cases

### DISCUSSION THREAD

**Discussion by osamakawish** [dotnet/csharplang#4393] [Language Ideas]
I'm not even sure why it's there... but I have examples of code that could be a whole lot cleaner if I could skip the break statement.

Any reason for why it's mandatory?

**Comment by HaloFour** [comment]
Yes, the `break` is required to avoid the common bug of accidentally falling through cases which occurs in C and Java.  If you want fallthrough you can use `goto` to jump to a specific case label.

**Reply by CyrusNajmabadi** [reply]
Halo's answer is correct.  It's intentional and desirable to prevent accidental fallthrough.  Note that we dont' require it when fallthrough is immediate and obvious.

**Reply by osamakawish** [reply]
> Yes, the `break` is required to avoid the common bug of accidentally falling through cases which occurs in C and Java. If you want fallthrough you can use `goto` to jump to a specific case label.

Will that necessarily execute the `goto` though? Or will it only execute the `goto` case if it actually satisfies the case?

As in, is this:
```c#
switch (i) {
    case A: DoA(); goto case B;
    case B: DoB();
}
```
equivalent to:
```c#
if (i == A) { DoA(); DoB(); }
else if (i == B) { DoB(); }
```
or is it equivalent to:
```c#
if (i == A) { DoA(); }
if (i == B) { DoB(); }
```
Because that last result is what I want. Which is what a switch with no breaks would generally do.

**Reply by sharwell** [reply]
Technically we could remove the requirement and leave analysis for an analyzer. However, if we know that removal will be followed immediately by a built-in analyzer that is enabled by default, it doesn't seem worth the effort to go through a language change process.

With that said, it's interesting that the language opted for an error in this case. Most situations involving unambiguous semantics use warnings instead of errors to suggest better practices.

**Reply by osamakawish** [reply]
I think a different concept to a switch would've been a better alternative here, like a switch that breaks by default rather than continuing by default. The original switch is still useful in some cases. Maybe it could be called a `flip` or something.

**Reply by HaloFour** [reply]
```cs
int i = 1;

switch (i) {
    case 1: Console.WriteLine("Hello"); goto case 2;
    case 2: Console.WriteLine("World"); break;
}
```

Will print:

```
Hello
World
```

[SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAgrgOwD4AEBMAjALABQZZAlpjAASW0C8thA3BaVAO6UwDGAC1oAKSgEpaAbzK1ZtPtDgsQtXIQCcwgEQAJOABt9Aey1jWtAOZGYR+Ytr52pOXahL8KtZq0B1Iwn0AE1NzYAQ4CABrJwBfIA==)

> Which is what a switch with no breaks would generally do.

That's not what a `switch` without `break` does in C or Java.  The execution will implicitly fallthrough to the statements under the following case.

**Reply by osamakawish** [reply]
Sorry, I think my example was a bit oversimplified.

Say you have an enum with multiple flags that you're switching over `[Flags] enum ABC { A, B, C }`.

What I need to do is for the switch to follow through to each one of those cases, and see if the variable in the switch contains multiple of those flags.

That is, I want:

```c#
if ((i & A) != 0) DoA();
if ((i & B) != 0) DoB();
```

whereas the `goto` seems to do:

```c#
if ((i & A) != 0) {DoA(); DoB();}
if ((i & B) != 0) {DoB();}
```

**Reply by CyrusNajmabadi** [reply]
> I think a different concept to a switch would've been a better alternative here

We're not going to relitigate a language feature that was designed >20 years ago :)   C# switches follow their C heritage roots.  Changing this now is effectively a non-starter.

**Reply by CyrusNajmabadi** [reply]
> That is, I want:

It's unclear to me why use a switch then?  The `if` approach is clear and gives the semantics you need.

**Reply by osamakawish** [reply]
I don't think I'm really asking to redefine the switch statement though. Like in my last comment, I **want** it to follow through. I just need to make sure the `goto` case is satisfied before it's called, and I'm not sure if the `goto` actually does. In fact, it seems like it doesn't, I think...

**Reply by osamakawish** [reply]
> > That is, I want:
> 
> It's unclear to me why use a switch then? The `if` approach is clear and gives the semantics you need.

With the original switch without the forced `break` statement, that's essentially what the `switch` statement would do, would it not?

That is, this:

```c#
switch(i) {
    case A: DoA();
    case B: DoB();
}
```

is essentially the same as:

```c#
if ((i & A) != 0) DoA();
if ((i & B) != 0) DoB();
```

Is it not?

**Reply by CyrusNajmabadi** [reply]
> With the original switch without the forced break statement, that's essentially what the switch statement would do, would it not?

No, it is not.  Fallthrough doesn't retest the 'case'.  The 'case value' is only used for jumping.

**Reply by HaloFour** [reply]
And with `switch` expressions it becomes somewhat moot.

**Reply by CyrusNajmabadi** [reply]
> I want it to follow through

We're not going to do that.  It's considered a negative for fallthrough to ahppen automatically.  You can simulate fallthrough with an explicit `goto` that indicates to the compiler (and readers of the code) that this was intentional on your part.

**Reply by osamakawish** [reply]
> The 'case value' is only used for jumping.

I probably need to double check this. But if `(i & A) != 0` but `(i & B) == 0`, do you mean that the original switch case would compute both `DoA()` and `DoB()`? Despite the fact that `(i & B) == 0`?

**Reply by osamakawish** [reply]
Nvm... I think I misunderstood the original switch...

**Reply by HaloFour** [reply]
> Nvm... I think I misunderstood the original switch...

No worries, I think that's exactly why the original C# designers made the `break` mandatory.  They wanted to keep with the C syntax and behavior but prevent occurrences of fallthrough where the developer didn't intend for it to happen.  Looking back they definitely could have made the `break` implicit and achieved the same thing.

**Reply by CyrusNajmabadi** [reply]
> they definitely could have made the break implicit 

Yes. but then what would this mean:

```
switch (x)
{
case 0:
case 1:
   ...
```

if 'break' were implicit, then case 0 woudln't fallthrough to 1.  If we wanted that, we'd effectively be saying: break is implicit except for the empty case.  And i think that would have been more confusing.  How we have it today is how i prefer it.  You break when you break.  You fallthrough when tehre's no break.  Fallthrough is only allowed in the trivial case.

**Reply by HaloFour** [reply]
Yep, plus the argument that behaving differently than Java or C would have been confusing.

**Reply by HaloFour** [reply]
I also imagine that if the team had considered removing implicit fallthrough that they would have also considered different syntax for that scenario, such as `case 0, 1:`

**Comment by HaloFour** [comment]
Switch fallthrough never reevaluates the condition.  The switch jumps to the first case based on the expression and then executes all following statements until it hits a `break` or other control statement to leave the `switch` block.

Example in Java:
```java
  public class JavaFiddle
  {
    public static void main(String[] args)
    {
        int i = 1;
        
        switch (i) {
            case 1: System.out.println("Hello");
            case 2: System.out.println("World");
            default: System.out.println("Fallthrough");
        }
    }
  }
```
prints:
```
Hello
World
Fallthrough
```

Example in C:
```c
#include <stdio.h>

int main()
{
    int i = 1;
    
    switch (i) {
        case 1: printf("Hello\n");
        case 2: printf("World\n");
        default: printf("Fallthrough\n");
    }
}
```
prints:
```
Hello
World
Fallthrough
```

**Reply by canton7** [reply]
The confusion might be related to the fact that `case` statements which use patterns or `when` clauses *are* tested in order

**Comment by IS4Code** [comment]
It remains to specify whether omitting the `break` means fallthrough or implicit `break`. I have nothing against implicit `break`, but it may cause confusion when compared with other C-like languages. It might also be inconsistent with multiple consecutive `case` labels, as this is one of the occasions where fallthrough is allowed.

**Reply by wes-sleeman** [reply]
It might be nice (and hopefully not lead to confusion) to allow implicit `break`s only for the last case. If that were downgraded from an error to a warning (to be honest, I wouldn't mind it being removed entirely, but that's the VB programmer in me), I don't think people would be too upset and it would certainly save those of us coming from other languages a bit of a headache if we didn't care. Can't speak for others, but needing the `break` at the end has always irked me.

**Comment by qrli** [comment]
In the early days, most C# programmers come from C/C++/Java, and a lot of code are ported from C/C++/Java. The strictness of `switch` helped to prevent common mistakes. But now, most new C# programmers have no C experience, and the `switch` legacy feels alien.

I wish `switch` expression to replace `switch` statement in 99% use cases. New comers don't need to know quirks of `switch` statement. E.g.  its fall-though behavior, as it is rarely used, even in legacy code.
